examples/avr/assert/main.cpp

```diff
@@ -0,0 +1,92 @@
+// coding: utf-8
+/* Copyright (c) 2016, Roboterclub Aachen e.V.
+ * All Rights Reserved.
+ *
+ * The file is part of the xpcc library and is released under the 3-clause BSD
+ * license. See the file `LICENSE` for the full license governing this code.
+ */
+// ----------------------------------------------------------------------------
+
+#include <xpcc/architecture/platform.hpp>
+#include <avr/pgmspace.h>
+
+using xpcc::accessor::asFlash;
+
+// Flash support on avr-gcc is so horribly broken.
+#define IFS(s) asFlash(PSTR(s))
+
+#define XPCC_CAN_MODULE_NAME 		"can"
+#define XPCC_IOBUFFER_MODULE_NAME 	"iobuffer"
+#define XPCC_UART_MODULE_NAME 		"uart"
+
+extern "C" void
+xpcc_abandon(const char * module,
+			 const char * location,
+			 const char * failure)
+{
+	serialStream << IFS("Assertion '")
+			<< asFlash(module) << '.'
+			<< asFlash(location) << '.'
+			<< asFlash(failure)
+			<< IFS("' failed! Abandoning.") << xpcc::endl;
+
+	Leds::setOutput();
+	while(1) {
+		Leds::write(1);
+		xpcc::delayMilliseconds(20);
+		Leds::write(0);
+		xpcc::delayMilliseconds(180);
+	}
+}
+
+static xpcc::Abandonment
+test_assertion_handler(const char * module,
+					   const char * /* location */,
+					   const char * /* failure */)
+{
+	serialStream << IFS("#1: '") << asFlash(module) << IFS("'!") << xpcc::endl;
+	// The strings are located in FLASH!!!
+	if (strcmp_P(module, PSTR(XPCC_IOBUFFER_MODULE_NAME)) == 0)
```

Comment:
This is a bit sucky, if your application should be portable between AVR and Cortex-M. But I see not other way than to place the strings into Flash, since this feature is supposed to be used on ATtinys too.